// @ui-kitten/components/ui/spinner/spinner.component.js
import React from 'react';
import { Animated, StyleSheet, View, } from 'react-native';
import { styled, } from '../../theme';
import { SpinnerAnimation, } from './animation';
import { Size } from '../measure/type';
/**
 * Styled `Spinner` component. Designed to be used as `ActivityIndicator` component
 *
 * @extends React.Component
 *
 * @property {boolean} animating - Determines whether component is animating. Default is `true`.
 *
 * @property {string} status - Determines the status of the component.
 * Can be `basic`, `primary`, `success`, `info`, `warning`, `danger` or `control`.
 * Default is `primary`.
 *
 * @property {string} size - Determines the size of the component.
 * Can be `tiny`, `small`, `medium`, `large`, or `giant`.
 * Default is `medium`.
 *
 * @overview-example SpinnerSimpleUsage
 *
 * @overview-example SpinnerSizes
 *
 * @overview-example SpinnerStatuses
 *
 * @example SpinnerDataLoading
 */
export class SpinnerComponent extends React.PureComponent {
    constructor() {
        super(...arguments);
        this.animation = new SpinnerAnimation(this.containerSize.height);
        this.startAnimation = () => {
            this.animation.start();
        };
        this.stopAnimation = () => {
            this.animation.stop();
        };
        this.getComponentStyle = (source) => {
            const start = {
                container: source.container,
                arc: source.start,
            };
            const end = {
                container: source.container,
                arc: source.end,
                overflow: { top: this.containerSize.height / 2 },
            };
            return { start, end };
        };
        this.renderArcElement = (style, size) => {
            const arcSize = new Size(size.width, size.height / 2);
            return (React.createElement(Animated.View, { style: [style.container, styles.absolute, size] },
                React.createElement(View, { style: [styles.noOverflow, style.overflow, arcSize] },
                    React.createElement(Animated.View, { style: [style.arc, size] },
                        React.createElement(View, { style: [styles.noOverflow, arcSize] },
                            React.createElement(View, { style: [this.props.themedStyle, this.props.style] }))))));
        };
    }
    get containerSize() {
        const { width, height } = StyleSheet.flatten([this.props.themedStyle, this.props.style]);
        // @ts-ignore: width and height are restricted to be a number
        return new Size(width, height);
    }
    componentDidMount() {
        if (this.props.animating) {
            this.startAnimation();
        }
    }
    componentDidUpdate(prevProps) {
        const animatingChanged = this.props.animating !== prevProps.animating;
        if (animatingChanged && this.props.animating) {
            this.startAnimation();
        }
        if (animatingChanged && !this.props.animating) {
            this.stopAnimation();
        }
    }
    componentWillUnmount() {
        this.animation.release();
    }
    render() {
        const containerSize = this.containerSize;
        const { start, end } = this.getComponentStyle(this.animation.toProps());
        return (React.createElement(View, { style: containerSize },
            this.renderArcElement(start, containerSize),
            this.renderArcElement(end, containerSize)));
    }
}
SpinnerComponent.styledComponentName = 'Spinner';
SpinnerComponent.defaultProps = {
    animating: true,
};
const styles = StyleSheet.create({
    absolute: StyleSheet.absoluteFillObject,
    noOverflow: {
        overflow: 'hidden',
    },
});
export const Spinner = styled(SpinnerComponent);
