// @ui-kitten/components/ui/animation/animation.js
export const DEFAULT_CONFIG = {
    cycles: 1,
    useNativeDriver: true,
};
export class Animation {
    constructor(config) {
        this.counter = 0;
        this.running = false;
        this.onAnimationEnd = (result) => {
            this.counter += 1;
            if (this.counter === this.config.cycles) {
                this.stop();
            }
            if (this.running) {
                this.start(this.endCallback);
            }
            if (!this.running) {
                this.counter = 0;
                this.endCallback && this.endCallback(result);
                this.endCallback = null;
            }
        };
        this.config = {
            ...DEFAULT_CONFIG,
            ...config,
        };
    }
    start(callback) {
        this.endCallback = callback;
        this.running = true;
        this.animation.start(this.onAnimationEnd);
    }
    stop() {
        this.running = false;
        this.animation.stop();
    }
    release() {
        this.stop();
    }
}
