// @ui-kitten/components/ui/animation/shakeAnimation.js
import { Animated, Easing, } from 'react-native';
import { Animation, } from './animation';
const DEFAULT_CONFIG = {
    start: 0.0,
    offset: 2.5,
    easing: Easing.linear,
    duration: 25,
    cycles: 8,
};
export class ShakeAnimation extends Animation {
    constructor(config) {
        super({ ...DEFAULT_CONFIG, ...config });
        this.value = new Animated.Value(this.config.start);
    }
    get animation() {
        const { start, offset, ...restConfig } = this.config;
        const startAnimation = Animated.timing(this.value, {
            toValue: this.counter % 2 !== 0 ? -offset : offset,
            ...restConfig,
        });
        const endAnimation = Animated.timing(this.value, {
            toValue: this.counter % 2 !== 0 ? offset : -offset,
            ...restConfig,
        });
        return Animated.sequence([
            startAnimation,
            endAnimation,
        ]);
    }
    toProps() {
        return {
            // @ts-ignore: Animated.Value is not assignable to a number, but it is a number
            style: {
                transform: [{ translateX: this.value }],
            },
        };
    }
}
