// @ui-kitten/components/ui/animation/zoomAnimation.js
import { Animated, Easing, } from 'react-native';
import { Animation, } from './animation';
const DEFAULT_CONFIG = {
    start: 1.0,
    end: 0.5,
    easing: Easing.linear,
    duration: 500,
};
export class ZoomAnimation extends Animation {
    constructor(config) {
        super({ ...DEFAULT_CONFIG, ...config });
        this.value = new Animated.Value(this.config.start);
    }
    get animation() {
        const { start, end, ...restConfig } = this.config;
        const startAnimation = Animated.timing(this.value, {
            toValue: end,
            ...restConfig,
        });
        const endAnimation = Animated.timing(this.value, {
            toValue: start,
            ...restConfig,
        });
        return Animated.sequence([
            startAnimation,
            endAnimation,
        ]);
    }
    toProps() {
        return {
            // @ts-ignore: Animated.Value is not assignable to a number, but it is a number
            style: {
                transform: [{ scale: this.value }],
            },
        };
    }
}
