// @ui-kitten/components/ui/menu/menuItem.component.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
import React from 'react';
import { Platform, StyleSheet, TouchableOpacity, View, } from 'react-native';
import { Interaction, styled, } from '../../theme';
import { Text, } from '../text/text.component';
import { allWithPrefix, WebEventResponder, } from '../support/services';
/**
 * `MenuItem` is a support component for `Menu`.
 *
 * @extends React.Component
 *
 * @property {string} title - Determines the title of the ListItem.
 *
 * @property {StyleProp<TextStyle>} titleStyle - Customizes title style.
 *
 * @property {(style: StyleType) => ReactElement} accessory - Determines the accessory of the component.
 *
 * @property {(style: ImageStyle) => ReactElement} icon - Determines the icon of the component.
 *
 * @property {MenuItemType[]} subItems - Determines the sub-items of the MenuItem.
 *
 * @property {(index: number, event: GestureResponderEvent) => void} onPress - Emits when component is pressed.
 *
 * @property {TouchableOpacityProps} ...TouchableOpacityProps - Any props applied to TouchableOpacity component.
 */
class MenuItemComponent extends React.Component {
    constructor() {
        super(...arguments);
        this.webEventResponder = WebEventResponder.create(this);
        this.onMouseEnter = () => {
            this.props.dispatch([Interaction.HOVER]);
        };
        this.onMouseLeave = () => {
            this.props.dispatch([]);
        };
        this.onFocus = () => {
            this.props.dispatch([Interaction.FOCUSED]);
        };
        this.onBlur = () => {
            this.props.dispatch([]);
        };
        this.onPress = (event) => {
            if (this.props.onPress) {
                this.props.onPress(this.props.menuIndex, event);
            }
        };
        this.onPressIn = (event) => {
            this.props.dispatch([Interaction.ACTIVE]);
            if (this.props.onPressIn) {
                this.props.onPressIn(this.props.menuIndex, event);
            }
        };
        this.onPressOut = (event) => {
            this.props.dispatch([]);
            if (this.props.onPressOut) {
                this.props.onPressOut(this.props.menuIndex, event);
            }
        };
        this.onLongPress = (event) => {
            if (this.props.onLongPress) {
                this.props.onLongPress(this.props.menuIndex, event);
            }
        };
        this.getComponentStyles = (style) => {
            const { paddingHorizontal, paddingVertical, backgroundColor } = style;
            const titleStyles = allWithPrefix(style, 'title');
            const indicatorStyles = allWithPrefix(style, 'indicator');
            const iconStyles = allWithPrefix(style, 'icon');
            return {
                container: {
                    paddingHorizontal: paddingHorizontal,
                    paddingVertical: paddingVertical,
                    backgroundColor: backgroundColor,
                },
                title: {
                    marginHorizontal: titleStyles.titleMarginHorizontal,
                    fontFamily: titleStyles.titleFontFamily,
                    fontSize: titleStyles.titleFontSize,
                    fontWeight: titleStyles.titleFontWeight,
                    lineHeight: titleStyles.titleLineHeight,
                    color: titleStyles.titleColor,
                },
                indicator: {
                    width: indicatorStyles.indicatorWidth,
                    backgroundColor: indicatorStyles.indicatorBackgroundColor,
                },
                icon: {
                    width: iconStyles.iconWidth,
                    height: iconStyles.iconHeight,
                    marginHorizontal: iconStyles.iconMarginHorizontal,
                    tintColor: iconStyles.iconTintColor,
                },
            };
        };
        this.renderIcon = (style) => {
            const iconElement = this.props.icon(style);
            return React.cloneElement(iconElement, {
                style: [style, iconElement.props.style],
            });
        };
        this.renderTitle = (style) => {
            const { title, titleStyle } = this.props;
            return (React.createElement(Text, { style: [style, titleStyle] }, title));
        };
        this.renderAccessory = (style) => {
            return this.props.accessory(style);
        };
        this.renderComponentChildren = (style) => {
            const { title, icon, accessory } = this.props;
            return [
                icon && this.renderIcon(style.icon),
                title && this.renderTitle(style.title),
                accessory && this.renderAccessory(style.icon),
            ];
        };
    }
    render() {
        const { themedStyle, style, ...restProps } = this.props;
        const { container, indicator, ...restStyles } = this.getComponentStyles(themedStyle);
        const [iconElement, textElement, accessoryElement] = this.renderComponentChildren(restStyles);
        return (React.createElement(TouchableOpacity, Object.assign({ activeOpacity: 1.0 }, restProps, this.webEventResponder.eventHandlers, { style: [styles.container, container, webStyles.container, style], onPress: this.onPress, onPressIn: this.onPressIn, onPressOut: this.onPressOut, onLongPress: this.onLongPress }),
            React.createElement(View, { style: [styles.indicator, indicator] }),
            React.createElement(View, { style: styles.subContainer },
                iconElement,
                textElement),
            accessoryElement));
    }
}
MenuItemComponent.styledComponentName = 'MenuItem';
const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'space-between',
    },
    subContainer: {
        flexDirection: 'row',
        alignItems: 'center',
    },
    indicator: {
        ...StyleSheet.absoluteFillObject,
        zIndex: 2,
    },
});
const webStyles = Platform.OS === 'web' && StyleSheet.create({
    container: {
        // @ts-ignore
        outlineWidth: 0,
    },
});
export const MenuItem = styled(MenuItemComponent);
