// @ui-kitten/components/ui/list/list.component.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
import React from 'react';
import { FlatList, StyleSheet, } from 'react-native';
import { styled, } from '../../theme';
/**
 * `List` component is a performant interface for rendering simple, flat lists. Extends `FlatList`. Renders list of
 * `ListItem` components or custom content.
 *
 * @extends React.Component
 *
 * @property {(info: ListRenderItemInfo<ItemT>, style: StyleType) => ReactElement} renderItem - Takes an
 * item from data and renders it into the list.
 *
 * @property {FlatListProps} ...FlatListProps - Any props applied to FlatList component.
 *
 * @overview-example ListSimpleUsage
 *
 * @overview-example ListCompositeItem
 *
 * @example ListInlineStyling
 * ```
 */
export class ListComponent extends React.Component {
    constructor() {
        super(...arguments);
        this.listRef = React.createRef();
        this.scrollToEnd = (params) => {
            this.listRef.current.scrollToEnd(params);
        };
        this.scrollToIndex = (params) => {
            this.listRef.current.scrollToIndex(params);
        };
        this.getComponentStyle = (source) => {
            return {
                container: source,
                item: {},
            };
        };
        this.getItemStyle = (source, index) => {
            const { item } = source;
            return item;
        };
        this.keyExtractor = (item, index) => {
            return index.toString();
        };
        this.renderItem = (info) => {
            const itemStyle = this.getItemStyle(this.props.themedStyle, info.index);
            const itemElement = this.props.renderItem(info, itemStyle);
            return React.cloneElement(itemElement, {
                style: [itemStyle, styles.item, itemElement.props.style],
                index: info.index,
            });
        };
    }
    scrollToOffset(params) {
        this.listRef.current.scrollToOffset(params);
    }
    render() {
        const { style, themedStyle, ...derivedProps } = this.props;
        const componentStyle = this.getComponentStyle(themedStyle);
        return (React.createElement(FlatList, Object.assign({ keyExtractor: this.keyExtractor }, derivedProps, { ref: this.listRef, style: [componentStyle.container, styles.container, style], renderItem: this.renderItem })));
    }
}
ListComponent.styledComponentName = 'List';
const styles = StyleSheet.create({
    container: {},
    item: {},
});
export const List = styled(ListComponent);
