// @ui-kitten/components/ui/popover/popover.component.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
import React from 'react';
import { StyleSheet, } from 'react-native';
import { ModalService } from '../../theme';
import { PopoverView, } from './popoverView.component';
import { PopoverPlacementService } from './placement.service';
import { PlacementOptions, PopoverPlacements, } from './type';
import { MeasureElement, } from '../measure/measure.component';
import { Frame, Point, } from '../measure/type';
const POINT_OUTSCREEN = new Point(-999, -999);
/**
 * Displays content in a Modal positioned relative to child component.
 * Supports automatic positioning.
 *
 * @extends React.Component
 *
 * @method {() => void} show - Sets `content` element visible.
 *
 * @method {() => void} hide - Sets `content` element invisible.
 *
 * @property {boolean} visible - Determines whether popover is visible or not.
 *
 * @property {ReactElement} content - Determines the content of the popover.
 *
 * @property {ReactElement} children - Determines the element "above" which popover will be shown.
 *
 * @property {string | PopoverPlacement} placement - Determines the actualPlacement of the popover.
 * Can be `left`, `top`, `right`, `bottom`, `left start`, `left end`, `top start`, `top end`, `right start`,
 * `right end`, `bottom start` or `bottom end`.
 * Default is `bottom`.
 * Tip: use one of predefined placements instead of strings, e.g `PopoverPlacements.TOP`
 *
 * @property {boolean} fullWidth - Determines whether content element should have same width as child element.
 *
 * @property {StyleProp<ViewStyle>} backdropStyle - Determines the style of backdrop.
 *
 * @property {() => void} onBackdropPress - Determines component's behavior when the user is
 * tapping on back-drop.
 *
 * @property {ViewProps} ...ViewProps - Any props applied to View component.
 *
 * @overview-example PopoverSimpleUsage
 *
 * @overview-example PopoverPlacement
 *
 * @overview-example PopoverStyledBackdrop
 */
export class Popover extends React.Component {
    constructor() {
        super(...arguments);
        this.state = {
            childFrame: Frame.zero(),
            contentFrame: Frame.zero(),
            forceMeasure: false,
        };
        this.contentPosition = POINT_OUTSCREEN;
        this.placementService = new PopoverPlacementService();
        this.actualPlacement = this.preferredPlacement;
        this.show = () => {
            this.modalId = ModalService.show(this.renderMeasuringPopoverElement(), this.backdropConfig);
        };
        this.hide = () => {
            this.modalId = ModalService.hide(this.modalId);
        };
        this.onChildMeasure = (childFrame) => {
            this.state.childFrame = childFrame;
            if (!this.modalId && this.props.visible) {
                this.show();
                return;
            }
            if (this.modalId && this.props.visible) {
                ModalService.update(this.modalId, this.renderPopoverElement());
            }
        };
        this.onContentMeasure = (contentFrame) => {
            this.state.contentFrame = contentFrame;
            const placementOptions = this.findPlacementOptions(contentFrame, this.state.childFrame);
            this.actualPlacement = this.placementService.find(this.preferredPlacement, placementOptions);
            const displayFrame = this.actualPlacement.frame(placementOptions);
            this.contentPosition = displayFrame.origin;
            ModalService.update(this.modalId, this.renderPopoverElement());
        };
        this.findPlacementOptions = (contentFrame, childFrame) => {
            const width = this.props.fullWidth ? childFrame.size.width : contentFrame.size.width;
            const frame = new Frame(contentFrame.origin.x, contentFrame.origin.y, width, contentFrame.size.height);
            return new PlacementOptions(frame, childFrame, Frame.window(), Frame.zero());
        };
        this.renderContentElement = () => {
            const contentElement = this.props.content;
            const fullWidthStyle = { width: this.state.childFrame.size.width };
            return React.cloneElement(contentElement, {
                style: [this.props.fullWidth && fullWidthStyle, contentElement.props.style],
            });
        };
        this.renderPopoverElement = () => {
            return (React.createElement(PopoverView, Object.assign({}, this.props, { contentContainerStyle: [this.props.contentContainerStyle, styles.popoverView, this.contentFlexPosition], placement: this.actualPlacement.reverse() }), this.renderContentElement()));
        };
        this.renderMeasuringPopoverElement = () => {
            return (React.createElement(MeasureElement, { onMeasure: this.onContentMeasure }, this.renderPopoverElement()));
        };
    }
    get preferredPlacement() {
        return PopoverPlacements.parse(this.props.placement);
    }
    get contentFlexPosition() {
        const { x: left, y: top } = this.contentPosition;
        return { left, top };
    }
    get backdropConfig() {
        const { onBackdropPress, backdropStyle } = this.props;
        return { onBackdropPress, backdropStyle };
    }
    componentDidUpdate(prevProps) {
        if (!this.modalId && this.props.visible && !this.state.forceMeasure) {
            this.setState({ forceMeasure: true });
            return;
        }
        if (this.modalId && !this.props.visible) {
            this.contentPosition = POINT_OUTSCREEN;
            this.hide();
        }
    }
    componentWillUnmount() {
        this.hide();
    }
    render() {
        return (React.createElement(MeasureElement, { force: this.state.forceMeasure, onMeasure: this.onChildMeasure }, this.props.children));
    }
}
Popover.defaultProps = {
    placement: PopoverPlacements.BOTTOM,
};
const styles = StyleSheet.create({
    popoverView: {
        position: 'absolute',
    },
});
