// @ui-kitten/components/ui/popover/popoverIndicator.component.js
import React from 'react';
import { StyleSheet, View, } from 'react-native';
export class PopoverIndicator extends React.Component {
    constructor() {
        super(...arguments);
        this.getComponentStyle = (source) => {
            const flatStyle = StyleSheet.flatten(source);
            return {
                container: {
                    // @ts-ignore: `width` is restricted to be a number
                    borderLeftWidth: flatStyle.width,
                    // @ts-ignore: `width` is restricted to be a number
                    borderRightWidth: flatStyle.width,
                    // @ts-ignore: `height` is restricted to be a number
                    borderBottomWidth: flatStyle.height,
                    borderBottomColor: flatStyle.backgroundColor,
                    borderLeftColor: 'transparent',
                    borderRightColor: 'transparent',
                    backgroundColor: 'transparent',
                },
            };
        };
    }
    render() {
        const { style, ...props } = this.props;
        const componentStyle = this.getComponentStyle(style);
        return (React.createElement(View, Object.assign({}, props, { style: [style, styles.container, componentStyle.container] })));
    }
}
const styles = StyleSheet.create({
    container: {},
});
