// @ui-kitten/components/theme/modal/modalResolver.component.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
import React from 'react';
import { StyleSheet, TouchableOpacity, View, } from 'react-native';
export class ModalResolver extends React.Component {
    constructor() {
        super(...arguments);
        this.renderChildElement = (source) => {
            return React.cloneElement(source, {
                style: [source.props.style, this.props.style],
            });
        };
        this.renderComponentChildren = (source) => {
            return React.Children.map(source, this.renderChildElement);
        };
        this.renderComponent = () => {
            const componentChildren = this.renderComponentChildren(this.props.children);
            return (React.createElement(View, { style: StyleSheet.absoluteFill },
                React.createElement(TouchableOpacity, { style: [StyleSheet.absoluteFill, this.props.backdropStyle], activeOpacity: 1.0, onPress: this.props.onBackdropPress }),
                componentChildren));
        };
    }
    render() {
        return this.props.visible && this.renderComponent();
    }
}
ModalResolver.defaultProps = {
    visible: false,
};
