// @ui-kitten/components/theme/theme/withStyles.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
import React from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { ThemeContext } from './themeContext';
/**
 * `withStyles` is a High Order Function which is used to create themed style for non-styled component.
 * Basically used when need to use theme variable somewhere.
 * Returns component class which can be used as themed component.
 *
 * @property {ThemeType} theme - Determines theme used to style component.
 *
 * @property {Styles} themedStyle - Determines component style for it's current state.
 *
 * @param Component - Type: {ComponentType}. Determines class of component to be themed.
 *
 * @param createStyles - Type: {(theme: ThemeType) => Styles}. Determines arrow function used to create styles.
 *
 * @overview-example WithStylesSimpleUsage
 */
export const withStyles = (Component, createStyles) => {
    class Wrapper extends React.Component {
        constructor() {
            super(...arguments);
            this.withThemedProps = (props, theme) => {
                const themedStyle = createStyles && createStyles(theme);
                return { ...props, theme, themedStyle };
            };
            this.renderWrappedElement = (theme) => {
                const { forwardedRef, ...restProps } = this.props;
                const props = this.withThemedProps(restProps, theme);
                return (React.createElement(Component, Object.assign({}, props, { ref: forwardedRef })));
            };
        }
        render() {
            return (React.createElement(ThemeContext.Consumer, null, this.renderWrappedElement));
        }
    }
    const WrappingElement = (props, ref) => {
        return (
        // @ts-ignore
        React.createElement(Wrapper, Object.assign({}, props, { forwardedRef: ref })));
    };
    const ThemedComponent = React.forwardRef(WrappingElement);
    ThemedComponent.displayName = Component.displayName || Component.name;
    hoistNonReactStatics(ThemedComponent, Component);
    // @ts-ignore
    return ThemedComponent;
};
