// @ui-kitten/components/ui/bottomNavigation/bottomNavigationTab.component.js
/**
 * @license
 * Copyright Akveo. All Rights Reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */
import React from 'react';
import { Platform, StyleSheet, TouchableOpacity, } from 'react-native';
import { Interaction, styled, } from '../../theme';
import { Text, } from '../text/text.component';
import { isValidString, WebEventResponder, } from '../support/services';
/**
 * `BottomNavigationTab` component is a part of the `BottomNavigation` component.
 * `BottomNavigation` tabs should be wrapped in BottomNavigation to provide usable component.
 * See usage examples at `BottomNavigation` component documentation.
 *
 * @extends React.Component
 *
 * @property {boolean} selected - Determines whether component is selected.
 *
 * @property {string} title - Determines the title of the tab.
 *
 * @property {StyleProp<TextStyle>} titleStyle - Customizes title style.
 *
 * @property {(style: ImageStyle) => ReactElement} icon - Determines the icon of the tab.
 *
 * @property {(selected: boolean) => void} onSelect - Triggered on select value.
 *
 * @property {TouchableOpacityProps} ...TouchableOpacityProps - Any props applied to TouchableOpacity component.
 *
 * @overview-example BottomNavigationTabSimpleUsage
 *
 * @overview-example BottomNavigationTabWithIcon
 *
 * @example BottomNavigationTabInlineStyling
 */
export class BottomNavigationTabComponent extends React.Component {
    constructor() {
        super(...arguments);
        this.webEventResponder = WebEventResponder.create(this);
        // WebEventResponderCallbacks
        this.onMouseEnter = () => {
            this.props.dispatch([Interaction.HOVER]);
        };
        this.onMouseLeave = () => {
            this.props.dispatch([]);
        };
        this.onPress = () => {
            if (this.props.onSelect) {
                this.props.onSelect(!this.props.selected);
            }
        };
        this.getComponentStyle = (source) => {
            const { iconWidth, iconHeight, iconMarginVertical, iconTintColor, textMarginVertical, textFontFamily, textFontSize, textLineHeight, textFontWeight, textColor, ...containerParameters } = source;
            return {
                container: containerParameters,
                text: {
                    marginVertical: textMarginVertical,
                    fontFamily: textFontFamily,
                    fontSize: textFontSize,
                    lineHeight: textLineHeight,
                    fontWeight: textFontWeight,
                    color: textColor,
                },
                icon: {
                    width: iconWidth,
                    height: iconHeight,
                    marginVertical: iconMarginVertical,
                    tintColor: iconTintColor,
                },
            };
        };
        this.renderIconElement = (style) => {
            const iconElement = this.props.icon(style);
            return React.cloneElement(iconElement, {
                key: 1,
                style: [style, styles.icon, iconElement.props.style],
            });
        };
        this.renderTitleElement = (style) => {
            const { title, titleStyle } = this.props;
            return (React.createElement(Text, { key: 2, style: [style, styles.text, titleStyle] }, title));
        };
        this.renderComponentChildren = (style) => {
            const { icon, title } = this.props;
            return [
                icon && this.renderIconElement(style.icon),
                isValidString(title) && this.renderTitleElement(style.text),
            ];
        };
    }
    render() {
        const { style, themedStyle, ...restProps } = this.props;
        const { container, ...componentStyles } = this.getComponentStyle(themedStyle);
        const [iconElement, titleElement] = this.renderComponentChildren(componentStyles);
        return (React.createElement(TouchableOpacity, Object.assign({ activeOpacity: 1.0 }, restProps, this.webEventResponder.eventHandlers, { style: [container, styles.container, webStyles.container, style], onPress: this.onPress }),
            iconElement,
            titleElement));
    }
}
BottomNavigationTabComponent.styledComponentName = 'BottomNavigationTab';
const styles = StyleSheet.create({
    container: {
        justifyContent: 'center',
        alignItems: 'center',
    },
    text: {},
    icon: {},
});
const webStyles = Platform.OS === 'web' && StyleSheet.create({
    container: {
        // @ts-ignore
        outlineWidth: 0,
    },
});
export const BottomNavigationTab = styled(BottomNavigationTabComponent);
